### PR TITLE
Hardfork integration

### PR DESCRIPTION
**Pull request by sstanculeanu**
- integrated hardfork into new heartbeat v2 flow(added new channel for it with a TODO to remove the old one with the heartbeat v1 cleanup)
- moved `hardforkTrigger` into `processComponents` as it is needed on peerAuthentication interceptor
- refactored `processComponents` to properly create `hardforkTrigger` and `exportFactoryHandler`
- removed the option `WithHardforkTrigger` as this is available through `processComponents`
- covered the new `hardforkTrigger` flow into hardfork integration test

### CODE REVIEW

File: integrationTests/testProcessorNode.go [hunk 1]
```diff
@@ -2868,16 +2878,18 @@ func (tpn *TestProcessorNode) createHeartbeatWithHardforkTrigger(heartbeatPk str
 	processComponents.HistoryRepositoryInternal = tpn.HistoryRepository
 	processComponents.TxsSenderHandlerField = createTxsSender(tpn.ShardCoordinator, tpn.Messenger)
 
-	redundancyHandler := &mock.RedundancyHandlerStub{}
+	processComponents.HardforkTriggerField = tpn.HardforkTrigger
```
Comment (iulianpascalau) [inline integrationTests/testProcessorNode.go:2881]: :+1:

File: process/factory/interceptorscontainer/baseInterceptorsContainerFactory.go [hunk 2]
```diff
@@ -601,9 +607,12 @@ func (bicf *baseInterceptorsContainerFactory) generateUnsignedTxsInterceptors()
 func (bicf *baseInterceptorsContainerFactory) generatePeerAuthenticationInterceptor() error {
 	identifierPeerAuthentication := common.PeerAuthenticationTopic
 
+	internalMarshalizer := bicf.argInterceptorFactory.CoreComponents.InternalMarshalizer()
```
Comment (iulianpascalau) [inline process/factory/interceptorscontainer/baseInterceptorsContainerFactory.go:610]: internalMarshaller *

File: process/interceptors/processor/peerAuthenticationInterceptorProcessor.go [hunk 3]
```diff
@@ -52,11 +65,23 @@ func (paip *peerAuthenticationInterceptorProcessor) Validate(_ process.Intercept
 
 // Save will save the intercepted peer authentication inside the peer authentication cacher
 func (paip *peerAuthenticationInterceptorProcessor) Save(data process.InterceptedData, fromConnectedPeer core.PeerID, _ string) error {
-	interceptedPeerAuthenticationData, ok := data.(interceptedDataMessageHandler)
+	interceptedPeerAuthenticationData, ok := data.(interceptedPeerAuthenticationMessageHandler)
 	if !ok {
 		return process.ErrWrongTypeAssertion
 	}
 
+	payloadBuff := interceptedPeerAuthenticationData.Payload()
```
Comment (iulianpascalau) [inline process/interceptors/processor/peerAuthenticationInterceptorProcessor.go:73]: by checking the hardfork payload here means that the hardfork public key should be a registered validator, which is not the case. We need an exception for the intercepted peer auth message coming the registered public key from the config.
That is why I have left the TODO on L211 in process/heartbeat/interceptedPeerAuthentication.go

File: factory/heartbeatV2Components.go [hunk 4]
```diff
@@ -109,6 +110,12 @@ func (hcf *heartbeatV2ComponentsFactory) Create() (*heartbeatV2Components, error
 		}
 	}
 
+	hardforkPubKey := hcf.config.Hardfork.PublicKeyToListenFrom
```
Comment (iulianpascalau) [inline factory/heartbeatV2Components.go:113]: we have the same pattern just about everywhere. Wondering if we could convert & store the hardforkPubKeyBytes in the core components

File: integrationTests/testProcessorNode.go [hunk 5]
```diff
@@ -2888,7 +2893,7 @@ func (tpn *TestProcessorNode) createHeartbeatWithHardforkTrigger() {
 
 	// TODO: remove it with heartbeat v1 cleanup
 	// =============== Heartbeat ============== //
-	redundancyHandler := &mock.RedundancyHandlerStub{}
+	/*redundancyHandler := &mock.RedundancyHandlerStub{}
```
Comment (iulianpascalau) [inline integrationTests/testProcessorNode.go:2896]: commented code?

File: factory/heartbeatV2Components.go (hunk 4 above)
Comment (sstanculeanu) [reply factory/heartbeatV2Components.go]: added, great idea :pray:

File: integrationTests/testProcessorNode.go (hunk 5 above)
Comment (sstanculeanu) [reply integrationTests/testProcessorNode.go]: my bad, forgot it while testing heartbeat v2 only..fixed it

File: factory/heartbeatV2Components.go (hunk 4 above)
Comment (sstanculeanu) [reply factory/heartbeatV2Components.go]: added, great idea :pray:

File: heartbeat/mock/hardforkTriggerStub.go [hunk 6]
```diff
@@ -4,13 +4,24 @@ import "github.com/ElrondNetwork/elrond-go/update"
 
 // HardforkTriggerStub -
 type HardforkTriggerStub struct {
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply heartbeat/mock/hardforkTriggerStub.go:6]: they were all the same..kept only one `HardforkTriggerStub` into testscommon and used it everywhere

File: integrationTests/testProcessorNode.go [hunk 7]
```diff
@@ -114,6 +114,8 @@ import (
 
 var zero = big.NewInt(0)
 
+var hardforkPubKey = "erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th"
```
Comment (iulianpascalau) [inline integrationTests/testProcessorNode.go:117]: this should have been a BLS public key. Can be resolved in a next PR

File: integrationTests/testProcessorNode.go [hunk 8]
```diff
@@ -2292,6 +2293,8 @@ func (tpn *TestProcessorNode) initNode() {
 	coreComponents.SyncTimerField = &mock.SyncTimerMock{}
 	coreComponents.EpochNotifierField = tpn.EpochNotifier
 	coreComponents.ArwenChangeLockerInternal = tpn.ArwenChangeLocker
+	hardforkPubKeyBytes, err := coreComponents.AddressPubKeyConverterField.Decode(hardforkPubKey)
```
Comment (iulianpascalau) [inline integrationTests/testProcessorNode.go:2296]: use coreComponents.ValidatorPubKeyConverterField instead

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**